eventsources/sources/kafka/start.go

```diff
@@ -166,3 +264,55 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 		}
 	}
 }
+
+// Consumer represents a Sarama consumer group consumer
+type Consumer struct {
+	ready            chan bool
+	dispatch         func([]byte) error
+	logger           *zap.SugaredLogger
+	kafkaEventSource *v1alpha1.KafkaEventSource
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (consumer *Consumer) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	close(consumer.ready)
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (consumer *Consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (consumer *Consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for message := range claim.Messages() {
+		//consumer.logger.Printf("Message claimed: value = %s, timestamp = %v, topic = %s", string(message.Value), message.Timestamp, message.Topic)
+		consumer.logger.Info("dispatching event on the data channel...")
+		eventData := &events.KafkaEventData{
+			Topic:     message.Topic,
+			Partition: int(message.Partition),
+			Timestamp: message.Timestamp.String(),
+			Metadata:  consumer.kafkaEventSource.Metadata,
+		}
+		if consumer.kafkaEventSource.JSONBody {
+			eventData.Body = (*json.RawMessage)(&message.Value)
+		} else {
+			eventData.Body = message.Value
+		}
+		eventBody, err := json.Marshal(eventData)
+		if err != nil {
+			consumer.logger.Desugar().Error("failed to marshal the event data, rejecting the event...", zap.Error(err))
+			continue
+		}
+
+		if err = consumer.dispatch(eventBody); err != nil {
+			consumer.logger.Desugar().Error("failed to dispatch kafka event...", zap.Error(err))
+		}
+
+		session.MarkMessage(message, "")
```

Comment:
Confirm - Is it expected to `MarkMessage` even `consumer.dispatch() fails`?

Reply:
Good catch fixed